### PR TITLE
Fill the customIdentifier in EC2

### PR DESCRIPTION
**Pull request by itchyny**
With this pull request, mackerel-agent on cloud platforms can identify the host by customIdentifier. Currently, EC2 is only supported.

Firstly, the agent on EC2 decides the customIdentifier based on the instance-id. Before creating a new host, the agent tries to find a host by the customIdentifier (if found, the host is possibly created by the AWS integration). On upgrading the version of mackerel-agent, it updates the host information with customIdentifier filled.

### CODE REVIEW

File: command/command.go [hunk 1]
```diff
@@ -56,13 +56,32 @@ func prepareHost(conf *config.Config, api *mackerel.API) (*mackerel.Host, error)
...
+				})
+				return filterErrorForRetry(lastErr)
+			})
+
+			if lastErr != nil {
+				return nil, fmt.Errorf("Failed to register this host: %s", lastErr.Error())
+			}
 		}
 
 		doRetry(func() error {
```
Comment (haya14busa) [inline command/command.go]: I guess we can move this block into  `if result == nil { ...} `

We have no need to retry to find host by id.

Comment (haya14busa) [inline command/command.go]: or, does it mean to check the host found by custom identifier has correct hostId? then, I want some comment.

File: mackerel/host.go [hunk 2]
```diff
@@ -10,10 +10,11 @@ type Host struct {
...
-	RoleFullnames []string               `json:"roleFullnames"`
-	Checks        []string               `json:"checks"`
-	DisplayName   string                 `json:"displayName,omitempty"`
+	Name             string                 `json:"name"`
+	Meta             map[string]interface{} `json:"meta"`
+	Interfaces       interface{}            `json:"interfaces"`
+	RoleFullnames    []string               `json:"roleFullnames"`
+	Checks           []string               `json:"checks"`
+	DisplayName      string                 `json:"displayName,omitempty"`
+	CustomIdentifier *string                `json:"customIdentifier,omitempty"`
```
Comment (haya14busa) [inline mackerel/host.go]: [just a comment?] I don't quite understand  using both string pointer and omitempty.

I guess just using omitempty and string type may be enough.
If we use both string pointer and omitempty, we may path string value.

https://play.golang.org/p/OZb7TXdXO1

File: spec/cloud.go [hunk 3]
```diff
@@ -153,6 +154,30 @@ func (g *EC2Generator) Generate() (interface{}, error) {
 	return results, nil
 }
 
+// SuggestCustomIdentifier suggests the identifier of the EC2 instance
+func (g *EC2Generator) SuggestCustomIdentifier() (*string, error) {
```
Comment (haya14busa) [inline spec/cloud.go]: it's not so much complicated code, but adding some simple test is preferable, i think

File: mackerel/host.go (hunk 2 above)
Comment (itchyny) [inline mackerel/host.go]: Thanks, you've got a point. I really dislike the code regarding the empty string as nothing so I wanted to return nil when the custom identifier is not available. But I didn't notice the behaviour of `omitempty` you pointed out, `string` is surely preferable in this case.

File: command/command.go (hunk 1 above)
Comment (itchyny) [inline command/command.go]: Good catch. I think I was too nervous...

File: command/command.go [hunk 4]
```diff
@@ -56,21 +56,40 @@ func prepareHost(conf *config.Config, api *mackerel.API) (*mackerel.Host, error)
...
-
-		if lastErr != nil {
-			return nil, fmt.Errorf("Failed to register this host: %s", lastErr.Error())
+		if customIdentifier != "" {
+			doRetry(func() error {
+				result, lastErr = api.FindHostByCustomIdentifier(customIdentifier)
+				return filterErrorForRetry(lastErr)
+			})
+			if result != nil {
+				hostID = result.ID
```
Comment (haya14busa) [inline command/command.go]: now, you can remove this code.

File: command/command.go [hunk 5]
```diff
@@ -56,21 +56,40 @@ func prepareHost(conf *config.Config, api *mackerel.API) (*mackerel.Host, error)
 	if hostID, err := conf.LoadHostID(); err != nil { // create
 		logger.Debugf("Registering new host on mackerel...")
```
Comment (haya14busa) [inline command/command.go]: maybe, this message should be after `if customIdentifier != "" {...}` block.

**Comment by itchyny** [comment]
Thanks for the review.

**Comment by yoheimuta** [comment]
[MEMO] We found a more (+1 min) delay to create a /var/lib/mackerel-agent/id file. This change has an impact for ec2 users who for example uses an auto scaling feature and mackerel apis. 

- Before (v0.32.2), the file is created and then cloud-init started. 
- Now (v0.33.0), cloud-init started and the file is created later.

```
2016/08/09 15:11:42 INFO <main> Starting mackerel-agent version:0.33.0, rev:f3a97ae, apibase:https://mackerel.io

# This process is retried for 60 seconds
## https://github.com/mackerelio/mackerel-agent/compare/v0.32.2...v0.33.0#diff-f2f57f6e81c1ea7ce925abd60673525cR59
2016/08/09 15:11:42 WARNING <command> No host was found for the custom identifier: i-d7597e48.ec2.amazonaws.com
...
2016/08/09 15:12:40 WARNING <command> No host was found for the custom identifier: i-d7597e48.ec2.amazonaws.com
2016/08/09 15:12:40 INFO <command> Start: apibase = https://mackerel.io, hostName = ip-10-0-1-207, hostID = 2LHbdNAA4yC
```

**Pull request merged**